libdnf/dnf-repo.c

```diff
@@ -1215,7 +1339,13 @@ dnf_repo_check_internal(DnfRepo *repo,
         if (!dnf_repo_set_timestamp_modified(repo, error))
             return FALSE;
         age_of_data =(g_get_real_time() - priv->timestamp_modified) / G_USEC_PER_SEC;
-        if (age_of_data > permissible_cache_age) {
+        metadata_expire = dnf_repo_get_metadata_expire(repo);
+        if(metadata_expire != 0)
+            valid_time_allowed = metadata_expire <= permissible_cache_age ? metadata_expire  : permissible_cache_age;
```

Comment:
Yeah...the way you've written it now, `dnf_context_set_cache_age()` overrides the metadata expire?  That makes sense to me as we definitely want a way for software to force a refresh.  rpm-ostree does indeed do stuff like:

https://github.com/projectatomic/rpm-ostree/blob/293479974194464f4217de4e13aae345dff3402d/src/app/rpmostree-compose-builtin-tree.c#L333

So your code as is looks good to me then.